### PR TITLE
Improve abstraction within classes related to bookmark transactions feature

### LINKED ISSUE #290: Abstract common code blocks
Currently, there are classes that use the same logic and hence utilise the similar blocks of code. Hence to improve abstraction within the code, these similar blocks of code should be abstracted out.

### LINKED ISSUE #292: Inconsistent message structure for tab specific bookmark related commands
Currently. the message structure for the bookmark transaction related commands that are tab specific are not consistent with the message structure for the transaction.

### LINKED ISSUE #293: In-app example for Add Bookmark Income doesn't make sense
The current error message when `addbi` is used incorrectly:
```
Invalid command format! 
add-bookmark-income: Adds a bookmark income to the finance tracker. Parameters: t/TITLE a/AMOUNT [c/CATEGORY]...
Example: add-bookmark-income t/Phone Bill a/24 c/Utilities.
```

`Phone Bill` is generally regarded as an expense and not income, which can be misleading to users.

### PR DESCRIPTION
**Pull request by siddarth2824**
Resolves #290.
Resolves #292, resolves #293.

Changes made:
* Abstracted out the parsing logic for `DeleteBookmarkExpenseCommandParser` and `DeleteBookmarkIncomeCommandParser` by creating a `DeleteBookmarkTransactionCommandParser` class.
*  Abstracted out the parsing logic for `EditBookmarkExpenseCommandParser` and `EditBookmarkIncomeCommandParser` by creating a `EditBookmarkTransactionCommandParser` class.
* Abstracted out the parsing logic for `ConvertBookmarkExpenseCommandParser` snd `ConvertBookmarkIncomeCommandParser` by creating a `ConvertBookmarkTransactionCommandParser` class.
* Abstracted out the main parsing logic for `AddBookmarkExpenseCommandParser` and `AddBookmarkIncomeCommandParser` by creating a `BookmarkTransactionBuilder` class which is a utility class that parses the input arguments and create a `BookmarkIncome` and `BookmarkExpense` respectively.

### CODE REVIEW

File: src/main/java/ay2021s1_cs2103_w16_3/finesse/logic/parser/bookmarkparsers/BookmarkTransactionBuilder.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    private BookmarkTransactionBuilder(Title title, Amount amount, Set<Category> categories) {
+        this.title = title;
+        this.amount = amount;
+        this.categories = categories;
+    }
+
+    /**
+     * Creates a {@code BookmarkTransactionBuilder} object from parsing the input arguments.
+     */
+    public static BookmarkTransactionBuilder parse(String args, String exceptionMessage)throws ParseException {
```
Comment (wltan) [inline src/main/java/ay2021s1_cs2103_w16_3/finesse/logic/parser/bookmarkparsers/BookmarkTransactionBuilder.java:40]:
```suggestion
    public static BookmarkTransactionBuilder parse(String args, String exceptionMessage) throws ParseException {
```

File: src/main/java/ay2021s1_cs2103_w16_3/finesse/logic/commands/bookmark/AddBookmarkIncomeCommand.java [hunk 2]
```diff
@@ -24,9 +24,9 @@
             + PREFIX_AMOUNT + "AMOUNT "
             + "[" + PREFIX_CATEGORY + "CATEGORY]...\n"
             + "Example: " + COMMAND_WORD + " "
-            + PREFIX_TITLE + "Phone Bill "
-            + PREFIX_AMOUNT + "24 "
-            + PREFIX_CATEGORY + "Utilities";
+            + PREFIX_TITLE + "Summer Internship "
+            + PREFIX_AMOUNT + "1000 "
+            + PREFIX_CATEGORY + "Work";
```
Comment (wltan) [inline src/main/java/ay2021s1_cs2103_w16_3/finesse/logic/commands/bookmark/AddBookmarkIncomeCommand.java:29]: File a bug report and link it in the PR description.

File: src/main/java/ay2021s1_cs2103_w16_3/finesse/logic/parser/bookmarkparsers/BookmarkTransactionBuilder.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        }
+
+        Title title = ParserUtil.parseTitleAndTrimBetweenWords(argMultimap.getValue(PREFIX_TITLE).get());
+        Amount amount = ParserUtil.parseAmount(argMultimap.getValue(PREFIX_AMOUNT).get());
+        Set<Category> categoryList = ParserUtil.parseCategories(argMultimap.getAllValues(PREFIX_CATEGORY));
+
+        return new BookmarkTransactionBuilder(title, amount, categoryList);
+    }
+
+
```
Comment (wltan) [inline src/main/java/ay2021s1_cs2103_w16_3/finesse/logic/parser/bookmarkparsers/BookmarkTransactionBuilder.java:71]:
```suggestion

```

**Review by wltan — CHANGES_REQUESTED** [review]
Looks like some of the command messages have changed. If this is intentional, file a bug report and link it here as a fix.

File: src/main/java/ay2021s1_cs2103_w16_3/finesse/logic/parser/FinanceTrackerParser.java [hunk 4]
```diff
@@ -172,9 +175,10 @@ public Command parseCommand(String userInput, UiState uiState) throws ParseExcep
         case ConvertBookmarkCommand.COMMAND_ALIAS:
             switch (uiCurrentTab) {
             case EXPENSES:
-                return new ConvertBookmarkExpenseCommandParser().parse(arguments);
+                return new ConvertBookmarkExpenseCommand(
+                        new ConvertBookmarkCommandParser().parse(arguments));
             case INCOME:
-                return new ConvertBookmarkIncomeCommandParser().parse(arguments);
+                return new ConvertBookmarkIncomeCommand(new ConvertBookmarkCommandParser().parse(arguments));
```
Comment (wltan) [inline src/main/java/ay2021s1_cs2103_w16_3/finesse/logic/parser/FinanceTrackerParser.java:181]: Should line 178-179 be a single line? Other commands are parsed in a single line too.

**Review by wltan — CHANGES_REQUESTED** [review]

**Review by wltan — APPROVED** [review]
LGTM!

**Review by zhaojj2209 — APPROVED** [review]
LGTM!

**Pull request merged**